[knot-dns] initial integration (take two)

jfoote opened the pull request:
Adds the initial integration for knot-dns. Fuzz targets build/execute as part of knot CI and use corpora from external repos.

This has been updated from https://github.com/google/oss-fuzz/pull/1049 to pass libfuzzer-related flags from the build environment (whoops!).

Dor1s commented on projects/knot-dns/Dockerfile:
```diff
@@ -0,0 +1,41 @@
+ gengetopt \
+ curl \
+ gperf \
+ wget \
```
Are all of these utilities needed? I don't see any use of curl, wget, etc

jfoote replied:
That may be a copy/paste error, will analyze and update as needed. Thanks.

Dor1s commented on projects/knot-dns/build.sh:
```diff
@@ -0,0 +1,81 @@
+./configure --enable-mini-gmp --enable-static --disable-shared --disable-documentation --prefix=$DEPS_PATH $NETTLE_CONFIGURE_FLAGS
+( make -j$(nproc) || make -j$(nproc) ) && make install
+if test $? != 0;then
+        echo "Failed to compile nettle"
```
I think this code is unreachable, since we use `bash -e` in the shebang line, but not a big deal

jfoote replied:
That makes sense, will fix

Dor1s commented on projects/knot-dns/build.sh:
```diff
@@ -0,0 +1,81 @@
+git submodule update --init -- ./packet_libfuzzer.in
+find ./packet_libfuzzer.in/ -type f -exec zip -u $OUT/packet_libfuzzer_seed_corpus.zip {} \;
+git submodule update --init -- ./zscanner_libfuzzer.in
+find ./zscanner_libfuzzer.in/ -type f -exec zip -u $OUT/packet_libfuzzer_seed_corpus.zip {} \;
```
I guess it should be `$OUT/zscanner_libfuzzer_seed_corpus.zip` ?

jfoote replied:
Another bug -- will fix. Thanks.

jfoote commented on projects/knot-dns/build.sh:
```diff
@@ -0,0 +1,81 @@
+make -j$(nproc)
+cd $SRC/knot-dns/tests-fuzz
+make check
+/bin/bash ../libtool   --mode=install /usr/bin/install -c packet_libfuzzer zscanner_libfuzzer '/out'
```
Understood; I believe this is an artifact of previously building separate targets for AFL- and libfuzzer- style fuzzer entry points.

jfoote commented:
@Dor1s Thank you for the review! I removed the unnecessary packages and fixed the corpus zip command. I did not change the name of the fuzz targets as that will require another PR to knot -- if you think it is strictly necessary just let me know and I'll make the change.

Dor1s commented:
@jfoote awesome, thanks for the changes! Regarding the binaries names, it's up to you, but it would be more painful if someone decides to rename fuzzers in the future (known crashes, stats, corpus will be broken -- all of that depends on the binary name). With that, I'd strongly recommend using more generic names from the beginning.

jfoote commented:
I've made a PR to knot with the changes to discuss with the team -- I'll post here when it settles (along with a commit if applicable)

jfoote commented:
@Dor1s We renamed the fuzz targets in knot-dns master and I updated the PR branch here accordingly

Dor1s commented:
Awesome!

Dor1s commented on projects/knot-dns/build.sh:
```diff
@@ -0,0 +1,77 @@
+make -j$(nproc)
+cd $SRC/knot-dns/tests-fuzz
+make check
+/bin/bash ../libtool   --mode=install /usr/bin/install -c fuzz_packet fuzz_zscanner '/out'
```
one more minor thing: please use `$OUT/` instead of `/out`

Dor1s commented:
Thanks @jfoote, merged!

jfoote commented:
Thanks!
